reactor-core/src/main/java/reactor/core/publisher/Flux.java

```diff
@@ -1745,6 +1745,45 @@
 				eager));
 	}
 
+
+	/**
+	 * Uses a resource, generated by a {@link Publisher} for each individual {@link Subscriber},
+	 * while streaming the values from a {@link Publisher} derived from the same resource.
+	 * Whenever the resulting sequence terminates, a provided {@link Function} generates
+	 * a "cleanup" {@link Publisher} that is invoked but doesn't change the content of the
+	 * main sequence. Instead it just defers the termination (unless it errors, in which case
+	 * the error suppresses the original termination signal).
+	 * <p>
+	 * Note that if the resource supplying {@link Publisher} emits more than one resource, the
+	 * subsequent resources are dropped ({@link Operators#onNextDropped(Object, Context)}). If
+	 * the publisher errors AFTER having emitted one resource, the error is also silently dropped
+	 * ({@link Operators#onErrorDropped(Throwable, Context)}).
+	 * An empty completion or error without at least one onNext signal triggers a short-circuit
+	 * of the main sequence with the same terminal signal (no resource is established, no
+	 * cleanup is invoked).
+	 *
+	 * <p>
+	 * <img class="marble" src="doc-files/marbles/usingWhenSuccessForFlux.svg" alt="">
+	 *
+	 * @param resourceSupplier a {@link Publisher} that "generates" the resource,
+	 * subscribed for each subscription to the main sequence
+	 * @param resourceClosure a factory to derive a {@link Publisher} from the supplied resource
+	 * @param asyncCleanup an asynchronous resource cleanup invoked when the resource
+	 * closure terminates (with onComplete, onError or cancel)
+	 * @param <T> the type of elements emitted by the resource closure, and thus the main sequence
+	 * @param <D> the type of the resource object
+	 * @return a new {@link Flux} built around a "transactional" resource, with asynchronous
+	 * cleanup on all terminations (onComplete, onError, cancel)
+	 * @deprecated prefer using the {@link #usingWhen(Publisher, Function, Function, BiFunction, Function)} version which is more explicit about all termination cases,
```

Comment:
funny enough, I find this one more preferable. When I was using this operator, I always needed 3 blocks:
1. create the resource
2. work with the resource
3. `finally` that will destroy the resource afterwards, no matter what happened (complete, error, cancel)

Since this operator is usually advertised as "async try-with-resources", it is essential to expect to have these 3 parameters.

Separation of complete/error/cancel will lead to the code duplication because (in my practice) in most of the cases the same code is invoked.

Just my 2c :)

Reply:
the callback themselves having error handling is something else entirely though... and a signal based approach would represent a larger change

Reply:
ok, what about this: leave the single callback version and the BiPredicate-with-3-callbacks versions undeprecated, and also introduce the single callback version to `Mono` for better API alignment. wdyt @bsideup ?

Reply:
`BiFunction` you mean? :)

I would say I'm fine with deprecating `usingWhen(Publisher,Function,Function,Function)` and `usingWhen(Publisher,Function,Function,Function,Function)`, but keeping `usingWhen(Publisher,Function,Function)` :)